client/relay-chain-rpc-interface/src/rpc_client.rs

```diff
@@ -0,0 +1,454 @@
+// Copyright 2022 Parity Technologies (UK) Ltd.
+// This file is part of Cumulus.
+
+// Cumulus is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Cumulus is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus.  If not, see <http://www.gnu.org/licenses/>.
+
+use backoff::{future::retry_notify, ExponentialBackoff};
+use cumulus_primitives_core::{
+	relay_chain::{
+		v2::{CommittedCandidateReceipt, OccupiedCoreAssumption, SessionIndex, ValidatorId},
+		Hash as PHash, Header as PHeader, InboundHrmpMessage,
+	},
+	InboundDownwardMessage, ParaId, PersistedValidationData,
+};
+use cumulus_relay_chain_interface::{RelayChainError, RelayChainResult};
+use futures::{
+	channel::mpsc::{Receiver, Sender},
+	StreamExt,
+};
+use jsonrpsee::{
+	core::{
+		client::{Client as JsonRPCClient, ClientT, Subscription, SubscriptionClientT},
+		Error as JsonRpseeError,
+	},
+	rpc_params,
+	types::ParamsSer,
+	ws_client::WsClientBuilder,
+};
+use parity_scale_codec::{Decode, Encode};
+use sc_client_api::StorageData;
+use sc_rpc_api::{state::ReadProof, system::Health};
+use sc_utils::mpsc::{tracing_unbounded, TracingUnboundedReceiver, TracingUnboundedSender};
+use sp_core::sp_std::collections::btree_map::BTreeMap;
+use sp_runtime::DeserializeOwned;
+use sp_storage::StorageKey;
+use std::sync::Arc;
+
+pub use url::Url;
+
+const LOG_TARGET: &str = "relay-chain-rpc-client";
+
+const NOTIFICATION_CHANNEL_SIZE_LIMIT: usize = 20;
+
+/// Client that maps RPC methods and deserializes results
+#[derive(Clone)]
+pub struct RelayChainRPCClient {
+	/// Websocket client to make calls
+	ws_client: Arc<JsonRPCClient>,
+
+	/// Retry strategy that should be used for requests and subscriptions
+	retry_strategy: ExponentialBackoff,
+
+	to_worker_channel: Option<TracingUnboundedSender<NotificationRegisterMessage>>,
+}
+
+/// Worker messages to register new notification listeners
+#[derive(Clone, Debug)]
+pub enum NotificationRegisterMessage {
+	RegisterBestHeadListener(Sender<PHeader>),
+	RegisterImportListener(Sender<PHeader>),
+	RegisterFinalizationListener(Sender<PHeader>),
+}
+
+/// Worker that should be used in combination with [`RelayChainRPCClient`]. Must be polled to distribute header notifications to listeners.
+pub struct RPCStreamWorker {
+	// Communication channel with the RPC client
+	client_receiver: TracingUnboundedReceiver<NotificationRegisterMessage>,
```

Comment:
I initially wanted to make it bounded, but the bounded sending requires `&mut self` which makes it super inconvenient in this case. Every stream getting method in the collator would need to take the mutable reference.

The channel is only used by the clients to request new notification streams, which are bounded. Also, the worker is started automatically and drains this channel when the client is constructed. The overall risk seems manageable to me in this case.

Reply:
> I initially wanted to make it bounded, but the bounded sending requires `&mut self` which makes it super inconvenient in this case. Every stream getting method in the collator would need to take the mutable reference.

FYI, you could always just use [the channel from `tokio`](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Sender.html) which doesn't require a `&mut self` to send. (:

Reply:
Very good suggestion, was not aware of this difference. Switching to tokio then to be safe.

Reply:
Sorry for being late to the party. Not sure @koute why the usage of `unbounded` for the service to worker communication makes you nervous. I mean, that some is some well defined communication. We also don't want to drop any of this communication. We also don't need any backpressure here, otherwise @skunert could also just make the methods take `&mut self`. 

TLDR: I don't see any reason why we should not use unbounded here. If the worker is dropped the channel is closed and if we don't poll the unbounded receiver anymore, there is something broken anyway.

Reply:
I think we should just default to bounded.

I just picked an unbounded one because the API was better for my usecase. The tokio channel solved this.

Reply:
Bounded max sense when you need back pressure. Here you don't need any back pressure.

Reply:
Sorry, it's just that I've seen enough code already that used an unbounded channels and was also well defined and looked innocuous but ended up causing problems. And a failure mode where it fails explicitly is easier to debug *if* something were to be broken here. Which is why if we'd end up in the "something's broken anyway" situation I'd prefer to have it alert us to that fact, instead of potentially just silently chugging along in the worst case.

This is just a "just in case" measure; as you say, yeah, we don't strictly *need* this here, but I also don't see why not to do this; in normal circumstances when everything works as it should (that is - the messages are handled fast enough that no backpressure will ever be triggered) it should essentially be functionally equivalent to an unbound channel.

That said, it's not a huge deal, so if you feel strongly about not having a bounded channel here I'm fine with reverting it. :P

Reply:
Okay okay, I "give" up 😂 but we should at least increase the channel size to 1000. If something is broken, we still will see it.

Reply:
I mean it probably makes no difference, but we use almost everywhere in our code the futures channel and futures select. That is something I don't want to die on, but yeah.

Reply:
Indeed. But I don't really see us ever dropping the `tokio` dependency, and if it's an integral part of `tokio` we might as well use it *if* needed. It's not like it's some random unmaintained crate fresh off crates.io.

I even once tried to compare whether switching from the futures' channel to tokio's channel would make a difference performance-wise (for a few particularly busy channels in substrate), but in that particular case I found the difference to be negligible. Or in other words, performance-wise they're comparable, so it shouldn't really matter which one we use *in general* (not that this aspect matters here; it doesn't), except `tokio`'s channels *do* have a richer API in those cases where we can make use of it.

Anyway, again, this is not a huge deal so from my point of view so I'm fine with whichever way @skunert wants to go.

Reply:
Lets stick to the tokio bounded channel in this case. Its more defensive here with no real downsides.